Integrate SGX DCAP quote verification into OE SDK (#3414)
openenclave/openenclave: docs/DesignDocs/SGX_QuoteVerify_Integration.md

Attestation: OE SDK Integration with Intel® SGX DCAP Quote Verification Library for SGX Evidence Verification
====

This design document proposes an update of the OE SDK implementation
for integration with the Intel® SGX Data Center Attestation Primitives (DCAP)
Quote Verification Library (QVL), for support of evidence verification in
ECDSA-p256 format.

# Motivation

> Existing OE SDK implements the functionality for SGX ECDSA quote verification.

Reviewer: (grammar nit)
```suggestion
The existing OE SDK implements the functionality for SGX ECDSA quote verification.
```

Author: Fixed.

As implemented in code file `common/sgx/quote.c`, the verification code
performs SGX cert chain verification and Enclave Identity

> verification, including x509 parser, cert revocation checking, JSON parser and

Reviewer: ```suggestion
verification, including X.509 parser, cert revocation checking, JSON parser,
```

Author: Fixed.

etc. Also it uses some 3rd-party libraries, such as mbedtls.

However, there are a few areas that could be improved:

> - Current SGX ECDSA quote verification logic in `common/sgx/quote.c`
>   is not a complete implementation, e.g.

Reviewer: ```suggestion
- The current SGX ECDSA quote verification logic in `common/sgx/quote.c`
  is not a complete implementation, e.g.:
```

Author: Fixed.

  - Only check quote version number in function `_validate_sgx_quote()`
    in file `common/sgx/quote.c`
  - Only allow TCB level `UpToDate`, all other TCB levels would be treated
    as invalid. But in SGX design, some other TCB levels such as `OutOfDate`,
    `SWConfigNeeded` should not be treated as critical error, it's up to

>     a verifier application to decide whehter the platform TCB is valid or not

Reviewer: ```suggestion
    a verifier application to decide whether the platform TCB is valid or not
```

Author: Fixed.

    for its intended usages.

> - In current implementaion, all the quote verification logic is

Reviewer: ```suggestion
- In the current implementation, all the quote verification logic is
```

Author: Fixed.

  built into verifier enclaves. This means that OE SDK greatly increases the TCB of
  verifier enclaves. If security bugs in verification logic,
  including 3rd-party component, are discovered, verifier enclaves need to be
  upgraded and rebuilt.

# User Experience

The proposed update only changes the internal implementation of the OE SDK
attestation software stack. It does not impact the
[OE SDK attestation API](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/Attestation_API_Proposal.md).

With the integration of Intel® SGX DCAP QVL (including an untrusted library
(called QVL when there is no ambiguity)
and a Quote Verification Enclave (QvE)), a verifier's call to OE SDK API
`oe_verify_evidence()` for verification of SGX ECDSA evidence would trigger
quote verification by invoking QVL or QvE, depending on whether the call is

> from the enclave side or the host side.

Reviewer: ```suggestion
from the host side or the enclave side, respectively.
```

Author: Fixed.

Integration of the DCAP QVL depends on the installation of
Intel® SGX DCAP packages `libsgx-dcap-quote-verify` and `libsgx-ae-qve` and
their dependencies, as well as proper configuration of the components and their
access to dependent backend services (e.g. Quote Provider Library and
Provisioning Certificate Cashing Service (PCCS)). Details for the DCAP quote
verification library installation and configuration are outside the scope of
this document.

# Specification

## Existing OE SDK Implementation

### Implementation of the SGX ECDSA-p256 verifier plugin

> The OE SDK framework implementation searchs for a verifier plugin that
> supports the requested evidence format, and invoke the `verify_evidence()` or

Reviewer: ```suggestion
The OE SDK framework implementation searches for a verifier plugin that
supports the requested evidence format, and invokes the `verify_evidence()` or
```

Author: Fixed.

`verify_report()` entry point of the selected plugin.

The SGX ECDSA-p256 verifier plugin is implemented in code files
`enclave/sgx/report.c`, `common/sgx/verifier.c` and
`common/sgx/quote.c`. The same source tree implements both the enclave-side and
host-side verifier plugins.

There are 2 different verifier application / enclave scenarios in current
implementation for SGX ECDSA quote verification:
- Scenario 1 - Call `oe_verify_report()` to verify SGX remote report (aka SGX quote)
- Scenario 2 - Call `oe_verify_evidence()` to verify evidence in format
  `OE_FORMAT_UUID_SGX_ECDSA_P256`.

Based on design doc [Remote Attestation Collaterals](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/RemoteAttestationCollaterals.md),
new API `oe_verify_evidence()` will supersede
`oe_verify_report()`. So we only discuss scenario 2 here. But in both scenarios,

> the same function `oe_verify_quote_with_sgx_endorsements()` is invoked for

Reviewer: ```suggestion
the same internal function `oe_verify_quote_with_sgx_endorsements()` is invoked for
```

Author: Fixed.

quote verification.

For SGX ECDSA-p256 quote verification, in the enclave-side and host-side plugin
implementation of function `_verify_evidence()` in `common/sgx/verifier.c`,
several functions are called.
- `oe_get_sgx_endorsements()` and `oe_parse_sgx_endorsements()`

>   - Get relevant endorsements, including SGX PCK cert CRL, TCB info, QE identity
>     and etc

Reviewer: ```suggestion
  - Get relevant endorsements, including SGX PCK cert CRL, TCB info, QE identity,
    etc.
```

Author: Fixed.

- `oe_verify_quote_with_sgx_endorsements()`
  - Verify quote with provided endorsements, including quote parsing, cert chain

>     veriifcation, TCB level matching and etc

Reviewer: ```suggestion
    verification, TCB level matching, etc.
```

Author: Fixed.

    const struct _sgx_ql_qve_collateral_t *p_quote_collateral,
    const time_t expiration_check_date,
    uint32_t *p_collateral_expiration_status,
    sgx_ql_qv_result_t *p_quote_verification_result,
    sgx_ql_qe_report_info_t *p_qve_report_info,
    uint32_t supplemental_data_size,
    uint8_t *p_supplemental_data)

```

Notes on the API function `sgx_qv_verify_quote()`:
- Verifier application (called the verifier) controls quote verification via trusted
  QvE or untrusted QVL by specifing parameter `p_qve_report_info`
  - Trusted QvE when this parameter is non-NULL, containing the `report`
    for the verifier enclave that verifies the QvE security properties:
    Quote verification would be done inside QvE, and QvE would return a report
    targeting the verifier enclave, it means the verifier can verify QvE's
    reutrn report and identity
  - Untrusted QVL when this parameter is NULL:
    Quote verification would be done inside untrusted QVL library, the verifier can
    use this way on a non-SGX capable system, but the result can not be
    cryptographically authenticated in this mode
- If verifier doesn't provide quote verification collateral (aka Endorsement),

>   QVL will try to load Quote Provide Library(QPL) and connect to 
>   PCK Cert Caching Sever(PCCS) to get corresponding verification collaterals

Reviewer: ```suggestion
  QVL will try to load the Quote Provide Library (QPL) and connect to 
  the PCK Cert Caching Server (PCCS) to get corresponding verification collaterals
```

Author: Fixed.

    sgx_cpu_svn_t tcb_cpusvn;       ///< CPUSVN of the remote platform's PCK Cert
    sgx_isv_svn_t tcb_pce_isvsvn;   ///< PCE_ISVNSVN of the remote platform's PCK Cert
    uint16_t pce_id;                ///< PCE_ID of the remote platform
    uint8_t sgx_type;               ///< Indicate the type of memory protection available on the platform, it should be
                                    ///< one of Standard (0) and Scalable (1)

    // Multi-Package PCK cert related flags, they are only relevant to PCK Certificates issued by PCK Platform CA
    uint8_t platform_instance_id[16];///< Value of Platform Instance ID, 16 bytes
    pck_cert_flag_enum_t dynamic_platform; ///< Indicate whether a platform can be extended with additional packages
                                           ///< via Package Add calls to SGX Registration Backend
    pck_cert_flag_enum_t cached_keys;      ///< Indicate whether platform root keys are cached by SGX Registration Backend
    pck_cert_flag_enum_t smt_enabled;      ///< Indicate whether a plat form has SMT (simultaneous multithreading) enabled

} sgx_ql_qv_supplemental_t;
```

## Proposed Changes

### Options for retrieving SGX endorsements

> There are two options for the OE SDK plugin library to retreive SGX endorsments.

Reviewer: ```suggestion
There are two options for the OE SDK plugin library to retrieve SGX endorsements.
```

Author: Fixed.

> #### Option 1: Keep existing implemention in SGX verifier plugin

Reviewer: ```suggestion
#### Option 1: Keep existing implementation in SGX verifier plugin
```

Author: Fixed.

> In the current plugin implementation, if the verifier doesn't provide SGX endoresements, it

Author: Fixed.

calls API `oe_get_sgx_endorsements()` to parse SGX ECDSA quote body to get PCK
cert chain first, then calls OCALL API `oe_get_quote_verification_collateral_ocall()`
to load Quote Provide Library(QPL) and connect to PCK Cert Caching Sever(PCCS)
to get corresponding verification collaterals, including CRL, TCB and QE

> Identity and etc.

Reviewer: ```suggestion
Identity, etc.
```

Author: Fixed.

#### Option 2: Ask DCAP QVL to retrieve endorsements

> If the verifier doesn't provide endorsments when calling DCAP QVL API, the QVL API

Author: Fixed.

implementation will parse quote body to get PCK cert chain, then load QPL
and connect to PCCS to get corresponding verification
collaterals.

#### Proposal: Ask DCAP QVL to retrieve endorsements
The proposal is to implement option 2, the reasons are:
- The Intel SGX DCAP QVL already has logic to parse quote, and gets
  verification collaterals
- The collateral definition may change from time to time, e.g. DCAP 1.7 release
  updated PCK cert extension to add new fields for support of multi-package
  platforms
- If we stay with the existing OE SDK implementaion
  - The implementation would need to be updated whenver the verification
    collateral defintion is changed
  - The OE SDK logic for retrieving and parsing endorsements increases
    verifier enclave TCB.

With option 2, OE SDK doesn't need to maintain the complex logic for SGX collateral
retrieving & parsing, nor need to be updated in response to changes
in SGX collateral definition.

### Link with the SGX DCAP QVL Library

> To align with the current implementaion of SGX quote-ex library integration,

Reviewer: ```suggestion
To align with the current implementation of SGX quote-ex library integration,
```

Author: Fixed.

the OE SDK host-side plugin library will dynamically detect the presence of QVL

> library and loads it at runtime. In case the QVL library can't be found,

Reviewer: ```suggestion
library and load it at runtime. In case the QVL library can't be found,
```

Author: Fixed.

```

### Update implementation of existing plugin functions

These plugin functions include `_verify_evidence()`, `_verify_report()`
and `oe_verify_quote_with_sgx_endorsements()`.

- Update API `oe_verify_quote_with_sgx_endorsements()` implementation as below:
  - Input parameter doesn't change.
  - In host-side plugin:
    - Set input parameter `p_qve_report_info` to NULL, then call host-side
      function `oe_sgx_verify_quote()`, which will call SGX QVL library to
      verify quote.
  - In enclave-side plugin:
    - Construct structure `sgx_ql_qe_report_info_t` as below, `nonce` and
      `app_enclave_target_info` are input, qve_report is QvE report which
      target to application enclave.
    ```C
    typedef struct _sgx_ql_qe_report_info_t {
      sgx_quote_nonce_t nonce;
      sgx_target_info_t app_enclave_target_info;
      sgx_report_t qve_report;
    }sgx_ql_qe_report_info_t;
    ```
    - Call OCALL function `oe_verify_quote_ocall()` to host-side plugin library
      to verify SGX quote.
    - With all the QvE return values, call new API.
      `oe_verify_qve_report_and_identity()` (described below) to
      verify QvE report and identity.

> - In API `_verify_evidence()`, remove function call `oe_get_sgx_endorsements()`,

Reviewer 2: Current implementation, oe_get_sgx_endorsements() is **not** called, if caller provides an endorsement buffer.

Author: Yes. The proposal is about ask Intel(R) QVL to retrieve endorsements.
- If caller provides endorsements, then it will pass the endorsements to QVL, then QVL/QvE will use caller provided endorsements to verify quote
- If caller doesn't provide endorsements, we will set the endorsements buffer to NULL, then QVL will try to retrieve endorsements automatically

So we don't need function `oe_get_sgx_endorsements()` in both cases.

  and only call `oe_parse_sgx_endorsements()` if caller provides an endorsement buffer

> - In API `_verify_report()` and relevant internal APIs, such as

Reviewer 2: Is this going to break oe_veriy_report()?

Author: It's not the same report in oe_verify_report(). This is the Intel(R) QvE output report, which target to user's application enclave, only user's application enclave can verify this report.

Reviewer 3: I think @yentsanglee's point is a little more general than the specific `_verify_report()` function call; the policy enforcement was implicitly integrated into our quote verification logic in the old codepaths, but the delegation of that responsibility to the QVL/QVE means that there is potentially extra work to maintain parity behavior with the old `oe_get_report()`. 

For example, the `__oe_sgx_set_minimum_crl_tcb_issue_date()` is still a valid public API (in bits) even though we hide it from the doxygen now. It would be helpful to have a  section that briefly talks about what behaviors change with `oe_verify_report()` if it just ends up using this converged code path with the `oe_get_evidence()` APIs.

  `oe_verify_report_internal()` and `oe_verify_sgx_quote()`, remove the code that
  loads QPL and retrieves endorsements.

### Add function in enclave-side plugin for QvE Identity verification

For enclave-side verifier plugin, after quote verification, it needs to verify that the result
comes from a trusted QvE.

Intel® SGX SDK provides a library named `sgx_dcap_tvl` to help the verifier to
verify QvE's identity, you can refer to [source file](https://github.com/intel/SGXDataCenterAttestationPrimitives/blob/master/QuoteVerification/dcap_tvl/sgx_dcap_tvl.cpp)

This library uses hardcoded QvE identity values, because:
- There is no need for x509 and JSON parser in verifier enclave's TCB.
- Most of QvE's identity attributes only change infrequently.

> The only identity info may change more often is QvE's ISV SVN, so the verifier needs

Reviewer 2: Isn't it QVL's responsibility for checking current QvE's product id, ISV SVN, etc against QvE Identity data from [QvE Identity at Intel® Provisioning
Cert Server](https://api.portal.trustedservices.intel.com/documentation#pcs-qve-identity-v2)? I am afraid hardcoded QvE identity values do not scale.

Author: Because QVL is an untrusted component, user cannot rely on QVL to verify QvE's identity. The good practice is to verify QvE's identity in user's enclave.

In the other word, an attacker can use faked QVL & QvE, if QvE's identity is checked by QVL, then user would have no way to find it's a faked QvE.

Reviewer 3: Just to clarify, this is just a description of the Intel implementation of the enclave side trusted library for using the QVE, and we're just porting it into OE, correct? (i.e. we're not changing the internal logic of `sgx_verify_qve_report_and_identity()` when we import it as `oe_verify_qve_report_and_identity()`)

Author: @CodeMonkeyLeet , yes, the proposal is about porting the trusted library into OE, and the internal logic would not changed.

to provide a SVN number as threshold, only when current QvE's ISV SVN is equal to
or larger than this threshold, the verifier can trust the QvE verification result.

> The verifier can get latest QvE ISV SVN from [QvE Identity at Intel® Provisioning

Reviewer 3: This seems to be a direct parallel to the QEIdentity, which we explicitly treat as part of the collateral bundle handled during attestation. Is there a reason why this isn't treated the same way?

Cert Server](https://api.portal.trustedservices.intel.com/documentation#pcs-qve-identity-v2)

As OE SDK cannot use Intel® SGX SDK trusted library directly, we need to port
this library into OE SDK as part of enclave-side plugin.

```C
oe_result_t oe_verify_qve_report_and_identity(
        const uint8_t *p_quote,
        uint32_t quote_size,
        const sgx_ql_qe_report_info_t *p_qve_report_info,
        time_t expiration_check_date,
        uint32_t collateral_expiration_status,
        sgx_ql_qv_result_t quote_verification_result,
        const uint8_t *p_supplemental_data,
        uint32_t supplemental_data_size,
        sgx_isv_svn_t qve_isvsvn_threshold)
```
All the data returned by the QvE are included in the QvE report data field for
integrity protection.
```C
QvE report_data = SHA256([nonce || quote || expiration_check_date ||

> expiration_status || verification_result || supplemental_data] || 32 - 0x00)

Author: - The hash in report data is used for guarantee the QvE output would NOT be changed during transition. All the items in 
   square brackets are QvE outputs, we will add the output hash into QvE report data. Then application enclave can 
   calculate a hash by using QvE outputs, then compare the hash with value in report data. In this way, application enclave  
   can make sure they are real QvE output. 
- SHA256 algorithm generates a fixed size 256-bit (32 bytes) hash, but SGX report data length is 64 bytes. So we will leave 
  the remaining 32 bytes as 0 in report data. 

I will add more description later.

```

The flow of the API function `oe_verify_qve_report_and_identity()` is as below:
- Verify QvE report.
- Verify report data (report data described as above).
- Check QvE Identity by comparing QvE report fields against hardcode values.
  - Check that Report.MRSIGNER equals to Hardcoded QvE MRSIGNER.
  - Check that Misc Select, Attribute and ProdID equal to the hardcoded values,
    Misc select and Attribute need to apply Mask before comparison.
- Check that Report.ISVSVN >= Hardcoded ISV SVN.

Options to provide QvE ISV SVN threshold:

> - Option 1: Hardcode QvE ISV SVN in enclave-side plugin. Everytime QvE

Reviewer: ```suggestion
- Option 1: Hardcode QvE ISV SVN in enclave-side plugin. Every time the QvE
```

Author: Fixed.

            ISV SVN is updated, the plugin library implementation will need to

>             be updated to stay in sync.

Reviewer 3: Do we have a strategy for keeping this logic in sync with the QVL/QVE releases? The ship vehicle is not explicitly mentioned (I expect it's part of the dcap_ql package) so it might be worth pointing out whether the package versioning dependency is separate from the QPL. 

I'll note that this recommendation seems to contradict an earlier desire for the QVL to handle endorsements so that OE doesn't have to keep in sync with the Intel changes to endorsements (which seem less frequent that updates to the libraries?)

Author: For the QvE identity verification trusted library, basically we don't need to keep sync with QVL/QvE release. Because nearly all the info in QvE identity would NOT change, except there is a critical issue, such as QvE signing key is leaked.  
But note that QvE ISV SVN is the only exception, it may change from time to time. 

Below QvE identity for example, you will see that except `isvsvn`, all other items are about QvE attribute and mrsigner.
```C
"miscselect":"00000000","miscselectMask":"FFFFFFFF","attributes":"01000000000000000000000000000000","attributesMask":"FBFFFFFFFFFFFFFF0000000000000000","mrsigner":"8C4F5775D796503E96137F77C68A829A0056AC8DED70140B081B094490C57BFF","isvprodid":2,"tcbLevels":[{"tcb":{"isvsvn":3},"tcbDate":"2019-12-11T00:00:00Z","tcbStatus":"UpToDate...
```

In addition, all other SGX endorsements, such as PCK Cert, CRL, TCB info, etc. The internal fields may change from time to time, that's why I propose to get rid of the complex parsing tasks in OE.

> - Option 2: Update existing enclave-side plugin API to allow the verifier to input

Author: Thanks. I think add an input in `oe_policy_t` is a possible solution, in this way, we don't need to change public API.
But it's only about Intel(R) QvE, so it's not TEE agnostic. 

Your option 3 still need to change public API or update `oe_policy_t`.

            QvE ISV SVN threshold value. But it requires plugin API change, also
            it's NOT TEE agnostic.

We propose option 1 to avoid plugin API change.

### Update claims list to add SGX Quote Verification status and QVL/QvE returned supplemental data
In SGX remote attestation, the verifier may want to provide a different quote
verification policy than the one enforced by the sgx_qv_verify_quote() API.

The proposal is to extend claims definition with new SGX related
claims.

SGX quote verification API `sgx_qv_verify_quote()` returns 2 types of errors,
fatal errors or non-fatal errors:
- Fatal error: such as cert chain verification failed, cert revoked, API will
  return an error code directly
- Non-fatal error, such as cert chain out-of-date, API will return successful, but
  there is an output parameter to indicate the specific error code. In this case,

>   the verifier can decide whehter the SGX quote is valid based on their own policy.

Reviewer: ```suggestion
  the verifier can decide whether the SGX quote is valid based on their own policy.
```

Author: Fixed.

We propose to introduce a new SGX-specific claim `OE_CLAIM_SGX_QUOTE_VERIFICATION_STATUS` to
indicate QVL/QvE return status. The verifier can check the status to see if there
is a non-fatal error during quote verification. This claim can have one of three values:

  - OE_OK - Quote verification passed, and TCB level is up to date

>   - Non-Fatal error code - API returns OE_OK,

Author: I agree with you. 
But @yentsanglee pointed out that some customers may not care about the non-fatal (warning) errors, maybe return OE_OK is a good choice, then only the customer who want have a strict verification policy, he/she can check the additional status in SGX claims.

>   - Non-Fatal error code - API returns OE_OK,

Reviewer 4: In current OE SDK API design, if a function returns some error code other than `OE_OK`, does it always abort and not return other results in output parameters?

>   - Non-Fatal error code - API returns OE_OK,

Reviewer 2: Current OE SDK API design, OE_OK is the only success return value. Attestation APIs follow the same pattern. My only concern is backward compatibility. Otherwise, I generally agree with @CodeMonkeyLeet and @shnwc that the caller can make its own judgement by all the returned value.

>   - Non-Fatal error code - API returns OE_OK,

Reviewer 3: To recap a point in one of the SIG-attestation meetings: OE SDK currently makes specific policy decisions to ignore some verification issues on behalf of the caller. These decisions align with Azure expectations due to how it manages its SGX configuration, but should not be the default in a general SDK intended for multi-cloud/provider use like OE moving forwards. In that regard, a break with backwards compatibility is necessary here, and we should just manage it as we would other breaking changes to minimize the impact to the current users.

                               // his/her enclave by linking LIV version SDK
   OE_SGX_CONFIG_AND_SW_HARDENING_NEEDED  // The TCB level of the platform is up
                                          // to date, but additional
                                          // configuration of the platform at
                                          // its current patching level may be
                                          // needed. Moreove, SGX SW Hardening
                                          // is also needed. For example,
                                          // attester may need enable/disable
                                          // SMT in BIOS, also he/she needs to
                                          // apply LVI mitgation in enclave

Note: SGX TCB = Platform CPU HW/FW TCB + SGX PCE TCB.
```
If the verifier needs to provide a different quote verification policy
beyond the policy enforced by the sgx_qv_verify_quote() API, the verifier can
request sgx_qv_verify_quote() API to return supplemental data. The detailed
supplemental data definition are described in above `Background` section.

> So we also propose to add SGX supplemental data fields the the list of claims.

Reviewer: ```suggestion
So we also propose to add SGX supplemental data fields to the list of claims.
```

Author: Fixed.

> [PR 3353](https://github.com/openenclave/openenclave/pull/3353) added SGX

Reviewer 2: There is at least one customer who is interested in raw TCB Info in JSON format. It may not be a hard requirement.

Author: May or May not?
If it's a hard requirement to return raw collateral data, then I need to consider the alternatives.

Reviewer 2: It may not. Let me get back to you on this.

Reviewer 4: One question would be what this customer wants the raw TCB info blob for? If their real requirements are the individual fields in this blob that we will provide as a set of claims, then this proposal would meet their requirements better.

Reviewer 2: It's a hard requirement but I understand in step 1 the raw TCB Info will be kept.

by himself if we only return the endorsements directly, but parsing x509 cert
chain/CRL and JSON is quite a complex task.

The new SGX-Specific claims definition (Pls refer to `Background` section for
the descripton of each SGX specific claim):

```C
#define OE_CLAIM_SGX_QUOTE_VERIFICATION_STATUS "sgx_quote_verification_status"

#define OE_CLAIM_SGX_TCB_LEVEL_DATE_TAG "sgx_tcb_level_date_tag"
#define OE_CLAIM_SGX_PCK_CRL_NUM "sgx_pck_crl_num"
#define OE_CLAIM_SGX_ROOT_CA_CRL_NUM "sgx_root_ca_crl_num"
#define OE_CLAIM_SGX_TCB_EVAL_REF_NUM "sgx_tcb_eval_ref_num"
#define OE_CLAIM_SGX_ROOT_KEY_ID "sgx_root_key_id"
#define OE_CLAIM_SGX_PCK_PPID "sgx_pck_ppid"
#define OE_CLAIM_SGX_TCB_CPUSVN "sgx_tcb_cpusvn"
#define OE_CLAIM_SGX_TCB_PCE_ISVSVN "sgx_tcb_pce_isvsvn"
#define OE_CLAIM_SGX_PCE_ID "sgx_pce_id"
#define OE_CLAIM_SGX_TYPE "sgx_type"
#define OE_CLAIM_SGX_PLATFORM_INSTANCE_ID "sgx_platform_instance_id"
#define OE_CLAIM_SGX_DYNAMIC_PLATFORM "sgx_dynamic_platform"
#define OE_CLAIM_SGX_CACHED_KEYS "sgx_cached_keys"
#define OE_CLAIM_SGX_SMT_ENABLED "sgx_smt_enabled"

#define OE_SGX_CLAIMS_COUNT 15
```

### Update `oe_sgx_extract_claims()` to remove SGX endorsement dependency

> As SGX endrosments will be retrieved by the SGX DCAP QVL library, the verifier

Reviewer: ```suggestion
As SGX endorsements will be retrieved by the SGX DCAP QVL library, the verifier
```

Author: Fixed.

    // Multi-Package PCK cert related flags, they are only relevant to PCK Certificates issued by PCK Platform CA
    uint8_t platform_instance_id[16];///< Value of Platform Instance ID, 16 bytes
    pck_cert_flag_enum_t dynamic_platform; ///< Indicate whether a platform can be extended with additional packages
                                           ///< via Package Add calls to SGX Registration Backend
    pck_cert_flag_enum_t cached_keys;      ///< Indicate whether platform root keys are cached by SGX Registration Backend
    pck_cert_flag_enum_t smt_enabled;      ///< Indicate whether a plat form has SMT (simultaneous multithreading) enabled

} sgx_ql_qv_supplemental_t;
```

## Proposed Changes

In order to smoothly switch quote verification from OE to SGX DCAP QVL/QvE, we
propose to implement the change in 2 steps as below:
- Step 1: Implement QVL-based quote verification and make it co-exist with the
  existing implementation
- Step 2: Remove existing implementation and makes the DCAP QVL a hard dependency

### Step 1: Implement QVL-based quote verification and make it co-exist with the existing implementation

> If SGX DCAP QVL/QvE are not installed, quote verification will continue to use existing

Reviewer 2: How do you uninstall quote verification on Windows?

Author: QVL/QvE are part of DCAP QL installer in Windows, so QVL/QvE would be uninstalled when user uninstall DCAP QL.

Reviewer 2: > QVL/QvE are part of DCAP QL installer in Windows, so QVL/QvE would be uninstalled when user uninstall DCAP QL.

Is DCAP QL also for quote generation? If so, the user cannot uninstall it.

Author: Yes, DCAP QL is for quote generation.

  - Try to call DCAP QVL to verify quote first
    - In host-side plugin:
      - Call host-side function `oe_sgx_verify_quote()` with SGX endorsements,
        which will call SGX QVL library to verify quote.
    - In enclave-side plugin:
      - Construct structure `sgx_ql_qe_report_info_t` as below, `nonce` and
        `app_enclave_target_info` are input, qve_report is QvE report which
        target to application enclave.
      ```C
      typedef struct _sgx_ql_qe_report_info_t {
        sgx_quote_nonce_t nonce;
        sgx_target_info_t app_enclave_target_info;
        sgx_report_t qve_report;
      }sgx_ql_qe_report_info_t;
      ```
      - Call OCALL function `oe_verify_quote_ocall()` to host-side plugin library
        to verify SGX quote.
      - With all the QvE return values, call new API.
        `oe_verify_qve_report_and_identity()` (described below) to
        verify QvE report and identity
  - If DCAP QVL returns error to indicate that QVL doesn't not exist on current
    platform, then automatically fallback to the existing implementation
  - Add log information to indicate whether the QVL-based or existing implementation
    is used

> - If QVL-based quote verification is used, in order to exactly align with current

Reviewer 2: Current implementation also accepts "SWHardeningNeeded" as a success case. See [tcbinfo.c](https://github.com/openenclave/openenclave/blob/master/common/sgx/tcbinfo.c#L243).

Author: Thanks, I will update the design doc to address this issue

Author: BTW, the comment in [tcbinfo.c](https://github.com/openenclave/openenclave/blob/master/common/sgx/tcbinfo.c#L243) is not accurate, only on the old platform which will be impacted by LVI, Intel will mark UpToDate tcb as SWHardeningNeeded

#define OE_CLAIM_SGX_DYNAMIC_PLATFORM "sgx_dynamic_platform"
#define OE_CLAIM_SGX_CACHED_KEYS "sgx_cached_keys"
#define OE_CLAIM_SGX_SMT_ENABLED "sgx_smt_enabled"

#define OE_SGX_CLAIMS_COUNT 15
```

#### Options for retrieving SGX endorsements when using QVL-based quote verification
There are two options for the OE SDK plugin library to retrieve SGX endorsements.

##### Option 1: Keep existing implementation in SGX verifier plugin
In the current plugin implementation, if the verifier doesn't provide SGX endorsements, it
calls API `oe_get_sgx_endorsements()` to parse SGX ECDSA quote body to get PCK
cert chain first, then calls OCALL API `oe_get_quote_verification_collateral_ocall()`
to load Quote Provide Library(QPL) and connect to PCK Cert Caching Sever(PCCS)
to get corresponding verification collaterals, including CRL, TCB and QE
Identity, etc.

##### Option 2: Ask DCAP QVL to retrieve endorsements
If the verifier doesn't provide endorsements when calling DCAP QVL API, the QVL API
implementation will parse quote body to get PCK cert chain, then load QPL
and connect to PCCS to get corresponding verification
collaterals.

> ##### Proposal: Ask DCAP QVL to retrieve endorsements

Reviewer 2: If QVL/QvE code path is chosen, the endorsements are going to be impacted. So this really is a decision to be made in step 1. Right?

Author: The endorsements change will be in step 2. In steps, it will pass endorsements from OE to QVL directly.

    try to detect & load the QVL library, return specific error if environment variable
    not set or QVL doesn't exist
  - Call QVL API to verify the quote

#### Add OCALL to support enclave-side SGX ECDSA-p256 quote verification
As quote verification will be done by DCAP QVL/QvE, for enclave-side plugin,
it requires OCALL to call host-side APIs.

In this proposal, we will add one OCALL `oe_verify_quote_ocall()` in
`edl/sgx/attestation.edl`.
The OCALL is used for passing quote buffer and related parameters to host side.
All other relevant logic will be implemented on the host side, in files
`host/sgx/ocalls.c` and `host/sgx/quote.c`.
If DCAP QVL cannot be found in host-side plugin, a specific error will be returned.

Note that currently only ECDSA-p256 quotes are supported, but the OCALL will keep
`format_id` and `opt_params` for forward compatibility.

```C
oe_result_t oe_verify_quote_ocall(
    [in] const oe_uuid_t* format_id,
    [in, size=opt_params_size] const void* opt_params,
    size_t opt_params_size,
    [in, size=quote_size] const void* p_quote,
    size_t quote_size,
    oe_datetime_t expiration_check_date,

>     [out] uint32_t *p_collateral_expiration_status,

Reviewer: This seems odd to have to do things in 2 passes, one to get untrusted values and then oe_verify_qve_report_and_identity() to verify those values are correct.  I didn't follow why this two-phase process was needed.

Author: > This seems odd to have to do things in 2 passes, one to get untrusted values and then oe_verify_qve_report_and_identity() to verify those values are correct. I didn't follow why this two-phase process was needed.

The `untrusted values` you mentioned are QvE output. When verifier uses QvE to verify the SGX quote, he/she needs to confirm 2 things.
- QvE is trusted and up to date
- The outputs are really come from QvE, and they have not been changed

So the 2nd phase is needed for above two items.

For enclave-side verifier plugin, after quote verification, it needs to verify that the result
comes from a trusted QvE.

Intel® SGX SDK provides a library named `sgx_dcap_tvl` to help the verifier to
verify QvE's identity, you can refer to [source file](https://github.com/intel/SGXDataCenterAttestationPrimitives/blob/master/QuoteVerification/dcap_tvl/sgx_dcap_tvl.cpp)

This library uses hardcoded QvE identity values, because:
- There is no need for x509 and JSON parser in verifier enclave's TCB.
- Most of QvE's identity attributes only change infrequently.

The only identity info may change more often is QvE's ISV SVN, so the verifier needs
to provide a SVN number as threshold, only when current QvE's ISV SVN is equal to
or larger than this threshold, the verifier can trust the QvE verification result.
The verifier can get latest QvE ISV SVN from [QvE Identity at Intel® Provisioning
Cert Server](https://api.portal.trustedservices.intel.com/documentation#pcs-qve-identity-v2)

As OE SDK cannot use Intel® SGX SDK trusted library directly, we need to port
this library into OE SDK as part of enclave-side plugin.

```C

> oe_result_t oe_verify_qve_report_and_identity(

Reviewer: To confirm: are you proposing this as an _internal_ API?

Author: > To confirm: are you proposing this as an _internal_ API?

Yes, this is an internal API used by verifier plugin to verify QvE identity.

The proposal is to implement option 2, the reasons are:
- The Intel SGX DCAP QVL already has logic to parse quote, and gets
  verification collaterals
- The collateral definition may change from time to time, e.g. DCAP 1.7 release
  updated PCK cert extension to add new fields for support of multi-package
  platforms
- If we stay with the existing OE SDK implementation
  - The implementation would need to be updated whenever the verification
    collateral definition is changed
  - The OE SDK logic for retrieving and parsing endorsements increases
    verifier enclave TCB.

With option 2, OE SDK doesn't need to maintain the complex logic for SGX collateral
retrieving & parsing, nor need to be updated in response to changes
in SGX collateral definition.

#### Update existing plugin functions to only use DCAP QVL/QvE
- In API `_verify_evidence()`, remove function call `oe_get_sgx_endorsements()`,
  and only call `oe_parse_sgx_endorsements()` if caller provides an endorsement buffer
- In API `_verify_report()` and relevant internal APIs, such as
  `oe_verify_report_internal()` and `oe_verify_sgx_quote()`, remove the code that
  loads QPL and retrieves endorsements

> - There is an API `__oe_sgx_set_minimum_crl_tcb_issue_date()`in header file

Reviewer: Just FYI, double-underscore shouldn't be used as a prefix (and this is noted in the OE developer guidelines), so at some point that function should be renamed.

Author: > Just FYI, double-underscore shouldn't be used as a prefix (and this is noted in the OE developer guidelines), so at some point that function should be renamed.

Note that this is an existing API in current implementation

## Proposed Changes

> In order to somoothly switch quote verification from OE to SGX DCAP QVL/QvE, we 

Reviewer 5: ```suggestion
In order to smoothly switch quote verification from OE to SGX DCAP QVL/QvE, we 
```

Author: Thanks, fixed.
